extern/sector-storage/sched.go

```diff
@@ -495,79 +531,200 @@ func (sh *scheduler) trySched() {
 	sh.openWindows = newOpenWindows
 }
 
-func (sh *scheduler) runWorker(wid WorkerID) {
-	var ready sync.WaitGroup
-	ready.Add(1)
-	defer ready.Wait()
+// context only used for startup
+func (sh *scheduler) runWorker(ctx context.Context, w Worker) error {
+	info, err := w.Info(ctx)
+	if err != nil {
+		return xerrors.Errorf("getting worker info: %w", err)
+	}
 
-	go func() {
-		sh.workersLk.RLock()
-		worker, found := sh.workers[wid]
-		sh.workersLk.RUnlock()
+	sessID, err := w.Session(ctx)
+	if err != nil {
+		return xerrors.Errorf("getting worker session: %w", err)
+	}
+	if sessID == ClosedWorkerID {
+		return xerrors.Errorf("worker already closed")
+	}
 
-		ready.Done()
+	worker := &workerHandle{
+		w:    w,
+		info: info,
 
-		if !found {
-			panic(fmt.Sprintf("worker %d not found", wid))
-		}
+		preparing: &activeResources{},
+		active:    &activeResources{},
+		enabled:   true,
+
+		closingMgr: make(chan struct{}),
+		closedMgr:  make(chan struct{}),
+	}
+
+	wid := WorkerID(sessID)
+
+	sh.workersLk.Lock()
+	_, exist := sh.workers[wid]
+	if exist {
+		// this is ok, we're already handling this worker in a different goroutine
```

Comment:
is this notable at least?

Reply:
I don't think this can happen outside of a test, potentially worth logging